### PR TITLE
systemd/systemd.go: support journald JSON messages with arrays for values

### PR DESCRIPTION
**Pull request by anonymouse64**
As per the man page,

Journal entries permit non-unique fields within the same log entry. JSON does not allow non-unique fields within objects. Due to this, if a non-unique field is encountered a JSON array is used as field value, listing all field values as elements.

Why journald thought it would be a good idea to support non-unique fields, who
knows, but we should handle this properly when we get messages from journald
this way.

Fixes: https://bugs.launchpad.net/snapd/+bug/1893279

CC @kyrofa

### CODE REVIEW

File: systemd/systemd.go [hunk 1]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
...
+	if err == nil {
+		return s, nil
+	}
+
+	// try runes next, this is the case if journald thinks the output is not
+	// safe ascii
+	r := []rune{}
+	err = json.Unmarshal(*valObject, &r)
+	if err == nil {
+		return string(r), nil
```
Comment (zyga) [inline systemd/systemd.go:728]: Python 3 rolls in its grave here ;-)

File: systemd/systemd.go [hunk 2]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
...
+	}
+
+	// finally try list of strings
+	stringSlice := []string{}
+	err = json.Unmarshal(*valObject, &stringSlice)
+	if err == nil {
+		// if the slice is of length 1, just promote it to a plain scalar string
+		if len(stringSlice) == 1 {
```
Comment (zyga) [inline systemd/systemd.go:736]: I'm uncertain this transformation is safe. Can you please justify it?

File: systemd/systemd.go [hunk 3]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
...
-	sus, ok := l["__REALTIME_TIMESTAMP"]
-	if !ok {
+	// since the __REALTIME_TIMESTAMP is underscored and thus "trusted" by
+	// systemd, we assume that it will always be a valid string and not try to
+	// handle any possible array cases
+	sus, err := l.parseLogRawMessageString("__REALTIME_TIMESTAMP", func([]string) (string, error) {
+		return "", errors.New("no timestamp")
+	})
+	if err != nil {
 		return time.Time{}, errors.New("no timestamp")
```
Comment (zyga) [inline systemd/systemd.go:755]: I think you can just return `err`, otherwise the returned error is lost and it was pretty much the same as the one you have here.

File: systemd/systemd.go [hunk 4]
```diff
@@ -709,28 +766,45 @@ func (l Log) Time() (time.Time, error) {
...
 func (l Log) Message() string {
-	if msg, ok := l["MESSAGE"]; ok {
-		return msg
+	// for MESSAGE, if there are multiple strings, just concatenate them with a
+	// newline to keep as much data from journald as possible
+	msg, err := l.parseLogRawMessageString("MESSAGE", func(stringSlice []string) (string, error) {
+		return strings.Join(stringSlice, "\n"), nil
+	})
+	if err != nil {
+		return "-"
```
Comment (zyga) [inline systemd/systemd.go:775]: Should we DEBUG log the error here, in case we have a bug in the parser somehow and people are trying to find out why?

**Review by zyga — APPROVED** [review]
LGTM with small remarks on error handling.

File: systemd/systemd.go [hunk 5]
```diff
@@ -246,8 +247,20 @@ type Systemd interface {
...
+// In almost all cases, the strings map to a single string value, but as per the
+// manpage for journalctl, under the json format,
+//
+//    Journal entries permit non-unique fields within the same log entry. JSON
+//    does not allow non-unique fields within objects. Due to this, if a
+//    non-unique field is encountered a JSON array is used as field value,
+//    listing all field values as elements.
+//
+// as such, we sometimes get array values which need to be handled differently,
+// and additionally if systemd thinks the value of a string is not safe ascii,
```
Comment (bboozzoo) [inline systemd/systemd.go:260]: Perhaps we could quote the bit from the manpage:

```
Fields containing non-printable or non-UTF8 bytes are
encoded as arrays containing the raw bytes individually
formatted as unsigned numbers.
```

File: systemd/systemd.go [hunk 6]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
...
+	// try runes next, this is the case if journald thinks the output is not
+	// safe ascii
+	r := []rune{}
+	err = json.Unmarshal(*valObject, &r)
+	if err == nil {
+		return string(r), nil
+	}
+
+	// finally try list of strings
+	stringSlice := []string{}
```
Comment (bboozzoo) [inline systemd/systemd.go:732]: Is there any evidence that indicates we may need to use [][]rune too?

File: systemd/systemd_test.go [hunk 7]
```diff
@@ -507,29 +508,113 @@ func (s *SystemdTestSuite) TestLogs(c *C) {
...
+		"MESSAGE": mustJSONMarshal("m1"),
+	}
+	l3 := Log{
+		// this is just "hello" in ascii
+		"MESSAGE": mustJSONMarshal([]rune{104, 101, 108, 108, 111}),
+	}
+	l4 := Log{
+		"MESSAGE": mustJSONMarshal(5),
+	}
+	c.Check(Log{}.Message(), Equals, "-")
```
Comment (bboozzoo) [inline systemd/systemd_test.go:572]: Can you add this entry too (`\r` at the end of the message from the bug report):
```
l5 := Log{ "MESSAGE": mustJSONMarshal([]rune{13})
c.Check(l5.Message(), Equals, "\r")
```

File: systemd/systemd.go [hunk 8]
```diff
@@ -709,28 +766,45 @@ func (l Log) Time() (time.Time, error) {
...
 // SID is the syslog identifier of the Log, if any; otherwise, "-".
 func (l Log) SID() string {
-	if sid, ok := l["SYSLOG_IDENTIFIER"]; ok {
-		return sid
+	// if there are multiple SYSLOG_IDENTIFIER values, just act like there was
+	// not one, making an arbitrary choice here is probably not helpful
+	sid, err := l.parseLogRawMessageString("SYSLOG_IDENTIFIER", func([]string) (string, error) {
+		return "", fmt.Errorf("multiple identifiers not supported")
+	})
+	if err != nil || sid == "" {
```
Comment (stolowski) [inline systemd/systemd.go:787]: Would it make sense to use some other tag (e.g. "*" instead of '-') to denote "multiple identifiers not supported" error case?

File: systemd/systemd.go [hunk 9]
```diff
@@ -709,28 +766,45 @@ func (l Log) Time() (time.Time, error) {
...
 
 // PID is the pid of the client pid, if any; otherwise, "-".
 func (l Log) PID() string {
-	if pid, ok := l["_PID"]; ok {
+	// look for _PID first as that is underscored and thus "trusted" from
+	// systemd, also don't support multiple arrays if we find then
+	pid, err := l.parseLogRawMessageString("_PID", func([]string) (string, error) {
+		return "", fmt.Errorf("multiple pids not supported")
+	})
+	if err == nil && pid != "" {
```
Comment (stolowski) [inline systemd/systemd.go:800]: Same question about denoting "multiple...." error case.

File: systemd/systemd_test.go [hunk 10]
```diff
@@ -507,29 +508,113 @@ func (s *SystemdTestSuite) TestLogs(c *C) {
...
+func mustJSONMarshal(v interface{}) *json.RawMessage {
+	b, err := json.Marshal(v)
+	if err != nil {
+		panic(fmt.Sprintf("couldn't marshal json in test fixture: %v", err))
+	}
+	msg := json.RawMessage(b)
+	return &msg
+}
+
+func (s *SystemdTestSuite) TestLogPIDWithNonTrivialKeyValues(c *C) {
```
Comment (stolowski) [inline systemd/systemd_test.go:521]: Nice test

File: systemd/systemd.go [hunk 11]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
 	return isTimeout
 }
 
+func (l Log) parseLogRawMessageString(key string, sliceHandler func([]string) (string, error)) (string, error) {
+	valObject, ok := l[key]
+	if !ok || valObject == nil {
+		// NOTE: journalctl says that sometimes if a json string would be too
+		// large null is returned, so we may miss a message here
+		return "", fmt.Errorf("key %q missing from message", key)
```
Comment (stolowski) [inline systemd/systemd.go:713]: Would it make sense to indicate such messages somehow in our output, e.g. "[message truncated]" or some such?

**Review by stolowski — COMMENTED** [review]
Looks good, a couple of wonderings.

File: systemd/systemd.go (hunk 2 above)
Comment (anonymouse64) [reply systemd/systemd.go:736]: if we get json like:

```json
{
  "MESSAGE": ["hello world"]
}
```

this code just bubbles it up as if it was actually
```json
{
  "MESSAGE": "hello world"
}

```

File: systemd/systemd.go (hunk 1 above)
Comment (anonymouse64) [reply systemd/systemd.go:728]: haha, well we are using runes so we should get some points there!

File: systemd/systemd.go (hunk 3 above)
Comment (anonymouse64) [reply systemd/systemd.go:755]: sure

File: systemd/systemd.go (hunk 4 above)
Comment (anonymouse64) [reply systemd/systemd.go:775]: I would like to import logger here to log a message, but we also have logging from the caller of this code in daemon, so I'm not sure we should be doing duplicated logging here, but I'm open to alternative proposals.

What I don't want to have happen is to have an application that issues lots of messages that we can't handle/transform here and flood snapd logs with useless messages about an issue with a message, because I imagine in an app finds itself sending a weird message, it has propensity to send _lots_ of those kinds of message.

File: systemd/systemd.go (hunk 6 above)
Comment (anonymouse64) [reply systemd/systemd.go:732]: I haven't seen evidence of it, but I also haven't dug into journald code deeply enough to know that it _can't_ happen, so sure let's just be extra safe and handle that here too :smile:

File: systemd/systemd_test.go (hunk 7 above)
Comment (anonymouse64) [reply systemd/systemd_test.go:572]: sure

File: systemd/systemd.go (hunk 11 above)
Comment (anonymouse64) [reply systemd/systemd.go:713]: yes perhaps we could distinguish between the `!ok` case and the `!valObject` case so we could return different errors

File: systemd/systemd.go (hunk 8 above)
Comment (anonymouse64) [reply systemd/systemd.go:787]: thinking about this again nearly 6 months later, this and the PID are only used in displaying the log message like this:

```
2021-01-07T20:27:06Z docker.dockerd[112562]: some message
```
where "docker.dockerd" is the SID.

in the case of no SID found we have:

```
2021-01-07T20:27:06Z -[112562]: some message
```

so I'm not sure what would look better here to use 

```
2021-01-07T20:27:06Z *[112562]: some message
```

and for the converse case of PID, we have:

```
2021-01-07T20:27:06Z docker.dockerd[-]: some message
```

and the proposal to use something else would look like:

```
2021-01-07T20:27:06Z docker.dockerd[*]: some message
```

as such I think it's best to just leave it as "-".

File: systemd/systemd.go (hunk 9 above)
Comment (anonymouse64) [reply systemd/systemd.go:800]: see other comment

File: systemd/systemd_test.go (hunk 7 above)
Comment (anonymouse64) [reply systemd/systemd_test.go:628]: done

File: systemd/systemd.go (hunk 6 above)
Comment (anonymouse64) [reply systemd/systemd.go:794]: added and tested now

File: systemd/systemd.go (hunk 11 above)
Comment (anonymouse64) [reply systemd/systemd.go:757]: message is now different for the `!ok` vs `!valObject` cases

File: systemd/systemd.go (hunk 5 above)
Comment (anonymouse64) [reply systemd/systemd.go]: added

File: systemd/systemd.go [hunk 12]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
...
+	s := ""
+	err := json.Unmarshal(*valObject, &s)
+	if err == nil {
+		return s, nil
+	}
+
+	// try runes next, this is the case if journald thinks the output is not
+	// safe ascii
+	r := []rune{}
+	err = json.Unmarshal(*valObject, &r)
```
Comment (bboozzoo) [inline systemd/systemd.go:775]: https://www.freedesktop.org/wiki/Software/systemd/json/ states:

> A field that contains non-printable or non-UTF8 is serialized as a number array instead. This is necessary to handle binary data in a safe way without losing data, since JSON cannot embed binary data natively. Each byte of the binary field will be mapped to its numeric value in the range 0..255.

If we have invalid utf8 here, we should hit the error in line 743 eventually.

File: systemd/systemd.go [hunk 13]
```diff
@@ -692,12 +705,56 @@ func IsTimeout(err error) bool {
...
+	if err == nil {
+		// if the slice is of length 1, just promote it to a plain scalar string
+		if len(stringSlice) == 1 {
+			return stringSlice[0], nil
+		}
+		// otherwise let the caller handle it
+		return sliceHandler(stringSlice)
+	}
+
+	return "", fmt.Errorf("failed to decode")
```
Comment (bboozzoo) [inline systemd/systemd.go:805]: Does not seem to be covered by tests, we probably need a test case for when a field is none of the string, []string, []rune, and one of invalid utf-8 since they explicitly mention it in the docs.

File: systemd/systemd.go (hunk 4 above)
Comment (bboozzoo) [reply systemd/systemd.go:837]: Perhaps we could use `- (error decoding original message)` to at least indicate a problem there.

File: systemd/systemd.go (hunk 8 above)
Comment (bboozzoo) [reply systemd/systemd.go:849]: FWIW journal does not print anything in such case:
```
import syslog
syslog.openlog("")
syslog.syslog("foo")
```
produces:
```
Jan 08 09:16:45 galeon [13860]: foo
```

File: systemd/systemd.go (hunk 12 above)
Comment (anonymouse64) [reply systemd/systemd.go]: I refactored to parse into a []byte, and then let Go figure out what to do about invalid UTF-8. I think it's fine to use Go's escaping here for invalid UTF-8 since this is used for `snap logs` which is essentially user-facing and so providing something that is readable is better than providing numbers etc. Users/programs concerned with accuracy or dealing with known binary data should probably be talking with journald directly instead of going through snapd like this.

File: systemd/systemd.go (hunk 13 above)
Comment (anonymouse64) [reply systemd/systemd.go]: Added more test cases for this sort of thing

File: systemd/systemd.go (hunk 4 above)
Comment (anonymouse64) [reply systemd/systemd.go]: I modified the error message returned here to include the error but only if the error was something other than the MESSAGE key missing.

File: systemd/systemd_test.go [hunk 14]
```diff
@@ -531,29 +532,178 @@ func (s *SystemdTestSuite) TestLogs(c *C) {
...
+		{
+			// this is "hel" and "lo" in ascii
+			mustJSONMarshal([][]rune{
+				{104, 101, 108},
+				{108, 111},
+			}),
+			"hel\nlo",
+			"arrays of rune arrays are converted to arrays of strings",
+		},
+		{
```
Comment (bboozzoo) [inline systemd/systemd_test.go:635]:
```suggestion
		{
			// this is invalid utf-8 string followed by a valid one
			mustJSONMarshal([][]byte{
				{65, 66, 67, 192, 69},
				{104, 101, 108, 108, 111},
			}),
			"ABC\xc0E\nhello",
			"arrays of bytes, some are invalid utf-8 strings",
		},
		{
```

**Review by bboozzoo — APPROVED** [review]
LGTM

**Review by stolowski — APPROVED** [review]
Thank you
